lib/pos_transaction_register.go

```diff
@@ -0,0 +1,655 @@
+	tr.txnMembership = make(map[BlockHash]*MempoolTx)
+	tr.totalTxnsSizeBytes = 0
+}
+
+// GetFeeTimeIterator returns an iterator over the transactions in the register. The iterator goes through all transactions
+// as ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeIterator() *FeeTimeIterator {
+	tr.RLock()
+	defer tr.RUnlock()
+
+	return &FeeTimeIterator{
+		bucketIterator:    tr.feeTimeBucketSet.Iterator(),
+		mempoolTxIterator: nil,
+	}
+}
+
+// GetFeeTimeTransactions returns all transactions in the register ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeTransactions() []*MempoolTx {
+	tr.RLock()
+	defer tr.RUnlock()
+
+	txns := []*MempoolTx{}
+	it := tr.GetFeeTimeIterator()
+	for it.Next() {
+		if txn, ok := it.Value(); ok {
+			txns = append(txns, txn)
+		}
+	}
+	return txns
+}
+
+// GetTransaction returns the transaction with the given hash if it exists in the register, or nil otherwise.
+func (tr *TransactionRegister) GetTransaction(hash *BlockHash) *MempoolTx {
+	if hash == nil {
+		return nil
+	}
+
+	tr.RLock()
+	defer tr.RUnlock()
+
+	return tr.txnMembership[*hash]
+}
+
+// PruneToSize removes transactions from the end of the register until the size of the register shrinks to the desired
+// number of bytes. The returned transactions, _prunedTxns, are ordered by lowest-to-highest priority, i.e. first
+// transaction will have the smallest fee, last transaction will have the highest fee. Returns _err = nil if no
+// transactions were pruned.
+func (tr *TransactionRegister) PruneToSize(maxSizeBytes uint64) (_prunedTxns []*MempoolTx, _err error) {
+	tr.Lock()
+	defer tr.Unlock()
+
+	// If the maximum number of bytes is greater or equal to the current size of the register, return.
+	if maxSizeBytes >= tr.totalTxnsSizeBytes {
+		return nil, nil
+	}
+
+	// If the register is empty, return.
+	if tr.feeTimeBucketSet.Empty() {
+		return nil, nil
+	}
+
+	// Determine how many bytes we need to prune and get the transactions to prune.
+	minPrunedBytes := tr.totalTxnsSizeBytes - maxSizeBytes
+	prunedTxns, err := tr.getTransactionsToPrune(minPrunedBytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "TransactionRegister.PruneToSize: Error getting transactions to prune")
+	}
+
+	// Remove the transactions from prunedTxns.
+	for _, txn := range prunedTxns {
+		if err := tr.removeTransactionNoLock(txn); err != nil {
+			return nil, errors.Wrapf(err, "TransactionRegister.PruneToSize: Error removing transaction %v", txn.Hash.String())
+		}
+	}
+	return prunedTxns, nil
+}
+
+func (tr *TransactionRegister) getTransactionsToPrune(minPrunedBytes uint64) (_prunedTxns []*MempoolTx, _err error) {
+	if minPrunedBytes == 0 {
+		return nil, nil
+	}
+
+	prunedBytes := uint64(0)
+	prunedTxns := []*MempoolTx{}
+
+	// Find the FeeTime bucket at the end of the Set. It'll have the smallest fee among the buckets in the register.
+	it := tr.feeTimeBucketSet.Iterator()
+	it.End()
```

Comment:
This it.End() thing is a little tricky. Add a comment explaining that you need to do it because you're doing a reverse iteration. Also add the same comment below.

Also why not something more like the following:
```
// We start with a call to it.End() to go *one past* the last item. This is required when iterating backward.
it := tr.feeTimeBucketSet.Iterator() && 
for it.End(); it.Prev(); {

}
```

Reply:
Added the comment. Eh, I tried using the iterator inline in the "for" but it end up looking more complicated.